File: Arboles/ts/kahn-lista-ady.cpp
#include <vector>
#include <map>
#include <string>
#include <cstdio>
#include <iostream>
#include <queue>

using namespace std;

int MAX = 10000;
int n, t = 0;
vector<vector<int>> adj(MAX);
vector<int> inc(MAX, 0);
vector<int> topo;

int main(){
  int m, u, v, i, j, vis = 0;
  queue<int> cola;
  cin >> n >> m;

  for(i = 0; i < n; ++i){
    cin >> u;
    adj[i].resize(u);
    for(j = 0; j < u; ++j){
      cin >> v;
      adj[i][j] = v;
      ++inc[v];
    }
  }

  cout << "Grafo" << endl;

  for(i = 0; i < n; ++i){
    cout << "Nodo " << i << ":";
    for(j = 0; j < adj[i].size(); ++j)
      cout << " " << adj[i][j];
    cout << endl;
  }

  cout << "Ordenamiento topológico" << endl;

  for(i = 0; i < n; ++i)
    if(inc[i] == 0)
      cola.push(i);

  while(!cola.empty()){
    v = cola.front();
    cola.pop();
    topo.push_back(v);

    for(i = 0; i < adj[v].size(); ++i){
      u = adj[v][i];
      --inc[u];
      if(inc[u] == 0)
	cola.push(u);
    }

    ++vis;
  }

  if(vis != n)
    printf("Hay un ciclo\n");
  else{
    for(i = 0; i < n; ++i)
      printf("%d%c", topo[i], (i == n - 1) ? '\n' : ' ');
  }

  return 0;
}

File: Arboles/ts/kahn-lista-ady.py
from sys import stdin
from collections import deque

MAX = 10000
adj = [[] for i in range(MAX)]
inc = [0 for i in range(MAX)]
n= int()
topo = []

def main():
    global n, topo
    n, m = list(map(int, stdin.readline().split()))
    vis = 0

    for i in range(n):
        vals = list(map(int, stdin.readline().split()))
        u = vals[0]
        for j in range(u):
            v = vals[j + 1]
            adj[i].append(v)
            inc[v] += 1

    print("Grafo")
    for i in range(n):
        print("Nodo %d:" % i)
        for j in range(len(adj[i])):
            print("%d" % adj[i][j], end = ' ')
        print("")

    print("Ordenamiento Topológico:")

    cola = deque()
    for i in range(n):
        if inc[i] == 0:
            cola.append(i)

    while len(cola) != 0:
        u = cola.popleft()
        topo.append(u)

        for i in range(len(adj[u])):
            v = adj[u][i]
            inc[v] -= 1
            if inc[v] == 0:
                cola.append(v)
        vis += 1

    if vis != n:
        print("Hay un ciclo")
    else:
        print(*topo)
        #for i in range(n):
            #print("%d" % topo[i], end = ' ')
    #print("")

main()

    
    
